internal/multiparty/manager.go

```diff
@@ -228,7 +228,34 @@ func (mm *multipartyManager) SubmitNetworkAction(ctx context.Context, signingKey
 	return err
 }
 
+func (mm *multipartyManager) prepareInvokeOperation(ctx context.Context, batch *core.BatchPersisted, contexts []*fftypes.Bytes32, payloadRef string) (*core.PreparedOperation, error) {
+	op, err := mm.txHelper.FindOperationInTransaction(ctx, batch.TX.ID, core.OpTypeBlockchainInvoke)
+	if err != nil || op == nil {
+		return nil, err
+	}
+	req, err := txcommon.RetrieveBlockchainInvokeInputs(ctx, op)
+	if err != nil {
+		return nil, err
+	}
+	return txcommon.OpBlockchainInvoke(op, req, &txcommon.BatchPinData{
+		Batch:      batch,
+		Contexts:   contexts,
+		PayloadRef: payloadRef,
+	}), nil
+}
+
 func (mm *multipartyManager) SubmitBatchPin(ctx context.Context, batch *core.BatchPersisted, contexts []*fftypes.Bytes32, payloadRef string) error {
+	if batch.TX.Type == core.TransactionTypeContractInvokePin {
+		preparedOp, err := mm.prepareInvokeOperation(ctx, batch, contexts, payloadRef)
+		if err != nil {
+			return err
+		} else if preparedOp != nil {
+			_, err = mm.operations.RunOperation(ctx, preparedOp)
```

Comment:
Biggest concern with the new functionality: since this is going to trigger a custom blockchain operation, there are a hundred ways it could fail. We're in an infinite retry loop here, and the assumption has always been that the blockchain piece cannot fail (which is pretty reasonable for the normal, very simple `pinBatch` contract).

If we have a fundamentally bad invoke here (wrong method, wrong params, wrong permissions, etc) then this processor will block forever with infinite retries. It will block any further batches of the same type from this author from being assembled.

In the case of private messages, the message has also claimed a nonce at this point, which means if the message never successfully gets pinned to the chain, that private group topic is also blocked.

Reply:
Agree with this concern. I believe this current "retry indefinitely" approach is safe from an existing functionality, and data integrity perspective. So I think merging this approach into main in this PR, then revisiting the design in a future PR ahead of a 1.3 release is the right way to go (rather than blocking this PR from merging)